common/morphology.F90

```diff
@@ -0,0 +1,217 @@
+!-------------------------------------------------------------------------------
+! Name: morphology.F90
+!
+! Purpose:
+! Module which provides simple morphological image transformation functions.
+!
+! Description and Algorithm details:
+! Four standard mathematical morphology functions are provided:
+! erode:  "Shrinks" brighter (more positive) objects in an image, by an amount
+!         defined by some kernel matrix (completely removing objects smaller
+!         than the kernel)
+! dilate: "Grows" brighter objects by an amount defined by a kernel matrix.
+! open:   Apply erode and dilate sequentially to the same image, using the same
+!         kernel. This will have the effect of removing any bright features
+!         smaller than the kernel - i.e. can be used to remove "salt-noise" or
+!         spikes in a image.
+! close:  The opposite of open. This has the effect of removing any dark
+!         features smaller than the kernel - i.e. can be used to remove
+!         "pepper-noise" or holes in an image.
+! All four functions accept two input arguments and return an integer matrix
+! with the same dimension as the first input array:
+! A(:,:) : The image to be transformed. Must be a 2-dimensional array (matrix)
+!          of integers. Note that binary data will work (i.e. bolean
+!          transformations will be done) but the output will still be integer
+!          type.
+! B(:,:) : The transform kernel to apply. This should also be a 2-dimensional
+!          integer array, smaller than A, with an odd number of elements in
+!          both dimensions.
+!
+! Arguments:
+! None (see above)
+!
+! History:
+! 2024/03/08, GT: First version
+!
+! Bugs:
+! None known
+!-------------------------------------------------------------------------------
+
+module morphology_m
```

Comment:
I note that the OpenCV library implements these functions. Someone should, one day, check how much more efficient they are than this.

Reply:
My version of this code in python implements the OpenCV versions and they're pretty fast. Can't do direct comparison as I never had the time but I suspect the runtime is, at worst, equal. From memory, the OpenCV ones are C# code wrapped in python.